### ISSUE TITLE
CRA/UglifyJS fails to minify common.js

### ISSUE THREAD

**Issue by d07RiV** [toji/gl-matrix#306]
When using the module in a CRA app, or any app that uses UglifyJS as a build step, I get the following message:

    Failed to minify the code from this file:

        ./node_modules/gl-matrix/src/gl-matrix/common.js:28

And links me to this:

https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#npm-run-build-fails-to-minify

Which, in turn, suggests that we file an issue and ask that the package be published pre-compiled.

**Comment by d07RiV** [comment]
Yes, they're suggesting ejecting and tweaking uglifyjs config, which is a bandaid fix for this problem. If I understand it correctly, it's complaining about the use of `let` keyword throughout the code, which could be fixed by publishing pre-compiled code instead.

**Comment by andrevenancio** [comment]
uglify has been updated to the latest version compatible with webpack 4 here https://github.com/toji/gl-matrix/pull/268

**Comment by stefnotch** [comment]
@d07RiV Is this still an issue?

**Comment by d07RiV** [comment]
The issue is with webpack being bundled with old version (0.4.6) of UglifyJS plugin, which is used by default in CRA and only supports ES5. I ejected and replaced it with the latest version of the plugin, which solved the issue, but I'd imagine it would still be a problem for anyone using the default, since it straight up can't handle ES6 syntax.

**Comment by andrevenancio** [comment]
Hi @d07RiV 
This issue has been fixed as far as I can tell. Can you provide an example code where this isn't working as expected? Also are you running `2.6.1`?

**Comment by d07RiV** [comment]
Still having the same issue.

1. Create a default create-react-app app
2. `npm install --save gl-matrix` (installs 2.6.1)
3. Add the following code (just to make sure the module is included in the build)

    import { vec3 } from 'gl-matrix';
    console.log(vec3.create());

4. `npm run build`

![image](https://user-images.githubusercontent.com/3448203/40869608-d955de9e-6626-11e8-990a-ba8b8229a3e1.png)

**Comment by stefnotch** [comment]
@d07RiV This issue probably has a workaround for your issue: https://github.com/toji/gl-matrix/issues/277

**Comment by andrevenancio** [comment]
Hey @d07RiV the idea of the "module" field on the package.json is to allow for webpack or rollup to go through the source and use tree shaking to remove any unused code that was imported.

this type of error makes me believe that you're missing a loader or rule on your webpack file and since you seem to be using `create-react-app` you should post this issue on that repo too.

**Comment by dtinth** [comment]
This issue is [documented in create-react-app’s README](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#npm-run-build-fails-to-minify) (emphasis mine):

> Some third-party packages don't compile their code to ES5 before publishing to npm. This often causes problems in the ecosystem because neither browsers (except for most modern versions) nor some tools currently support all ES6 features. **We recommend to publish code on npm as ES5 at least for a few more years.**
>
> To resolve this:
>
> 1. Open an issue on the dependency's issue tracker and ask that the package be published pre-compiled.
>
>    Note: Create React App can consume both CommonJS and ES modules. For Node.js compatibility, it is recommended that the main entry point is CommonJS. However, they can optionally provide an ES module entry point with the module field in package.json. **Note that even if a library provides an ES Modules version, it should still precompile other ES6 features to ES5 if it intends to support older browsers.**
>
> 2. Fork the package and publish a corrected version yourself.
>
> 3. If the dependency is small enough, copy it to your `src/` folder and treat it as application code.
>
> In the future, we might start automatically compiling incompatible third-party modules, but it is not currently supported. This approach would also slow down the production builds.

tl;dr:

- Compiling 3rd party libraries would make the bundling process slower, so `create-react-app` will not compile 3rd party libraries.

- They recommend module publishers to publish ES6 modules with no other ES6 features to remain compatible with ES5 browsers.

So IMO this issue comes down to whether `gl-matrix` would want to follow CRA’s advice or not.

**Comment by dtinth** [comment]
PR created: #335

**Comment by andrevenancio** [comment]
Hey! seems like a good idea. Will review your PR later to see it in action and then @stefnotch can have a look too cause I think he's the only one allowed to merge :)
